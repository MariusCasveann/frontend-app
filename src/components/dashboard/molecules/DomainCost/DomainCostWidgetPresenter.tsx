import React from 'react';
import * as chartjs from 'chart.js';
import { Doughnut } from 'react-chartjs-2';
import LoadingOrErrorPresenter from '../../../common/organisms/LoadingOrError/LoadingOrErrorPresenter';
import { hasChartData, hasNoChartData } from '../../helpers/checkChartData';
import { CardLegend } from '../../../common/molecules/CardLegend/CardLegend';
import { colors } from '../../../../config/colors';

interface DomainCostProps {
    loading: boolean;
    error: string;
    chartData?: chartjs.ChartData;
    formattedLabels: string[];
    messageWhenNoData?: string;
}

export default ({ chartData, formattedLabels, loading, error, messageWhenNoData }: DomainCostProps) => {
    if (hasChartData(chartData)) {
        return (
            <div className="domain-legend-container">
                {loading || error ? <LoadingOrErrorPresenter error={error} loading={loading} /> : false}
                {chartData && <CardLegend labels={formattedLabels} colors={colors} />}
                {chartData && (
                    <div className="canvas-container">
                        <Doughnut
                            data={chartData}
                            options={{
                                responsive: true,
                                legend: { display: false }
                            }}
                        />
                    </div>
                )}
            </div>
        );
    } else if (hasNoChartData(chartData)) {
        return <p>{messageWhenNoData}</p>;
    }

    return null;
};
