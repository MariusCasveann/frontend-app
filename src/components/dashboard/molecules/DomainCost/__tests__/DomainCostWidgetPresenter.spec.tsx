import * as React from 'react';
import renderer from 'react-test-renderer';
jest.mock('react-chartjs-2', () => ({ Doughnut: 'Doughnut' }));
jest.mock('../../../../common/organisms/LoadingOrError/LoadingOrErrorPresenter', () => 'LoadingOrErrorPresenter');
jest.mock('../../../../common/molecules/CardLegend/CardLegend', () => ({ CardLegend: 'CardLegend' }));
import DomainCostWidgetPresenter from '../DomainCostWidgetPresenter';

const mockData = {
    labels: [
        'Customer Retention(49198.94)',
        'Checkout (Multichannel Experience)(60463.31)',
        'B2B Experience(983.12)',
        'Service Management(6401.91)'
    ],
    datasets: [
        {
            data: [49198.94, 60463.31, 983.12, 6401.91],
            backgroundColor: ['rgb(109,46,235)', 'rgb(222,28,70)', 'rgb(220,159,172)', 'rgb(125,62,251)']
        }
    ]
};

const mockFormattedLabels = [
    'Assortment & Space Management (22412.97 EUR)',
    'B2B Experience (798.72 EUR)',
    'Checkout (Multichannel Experience) (48097.69 EUR)'
];

const mockEmptyData = {
    labels: [],
    datasets: [
        {
            data: [],
            backgroundColor: [],
            label: 'Domain costs for financial year"'
        }
    ]
};

describe('DomainCostWidgetPresenter', () => {
    describe('when there are chart data', () => {
        it('renders a Doughnut', () => {
            const wrapper = renderer
                .create(
                    <DomainCostWidgetPresenter
                        formattedLabels={mockFormattedLabels}
                        error=""
                        loading={false}
                        chartData={mockData}
                    />
                )
                .toJSON();
            expect(wrapper).toMatchSnapshot();
        });
    });

    describe('when there are no chart data', () => {
        it('doesnt render a Doughnut', () => {
            const wrapper = renderer
                .create(
                    <DomainCostWidgetPresenter
                        formattedLabels={mockFormattedLabels}
                        error=""
                        loading={false}
                        chartData={mockEmptyData}
                    />
                )
                .toJSON();
            expect(wrapper).toMatchSnapshot();
        });
    });

    describe('when chart data is undefined', () => {
        it('renders loading', () => {
            const wrapper = renderer
                .create(
                    <DomainCostWidgetPresenter
                        formattedLabels={mockFormattedLabels}
                        error={''}
                        loading={true}
                        chartData={undefined}
                    />
                )
                .toJSON();
            expect(wrapper).toMatchSnapshot();
        });
    });

    describe('when chart data is undefined', () => {
        it('renders an error', () => {
            const wrapper = renderer
                .create(
                    <DomainCostWidgetPresenter
                        formattedLabels={mockFormattedLabels}
                        error={'Error'}
                        loading={false}
                        chartData={undefined}
                    />
                )
                .toJSON();
            expect(wrapper).toMatchSnapshot();
        });
    });
});
