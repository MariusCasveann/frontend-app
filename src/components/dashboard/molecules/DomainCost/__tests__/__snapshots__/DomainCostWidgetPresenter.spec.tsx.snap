// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DomainCostWidgetPresenter when chart data is undefined renders an error 1`] = `null`;

exports[`DomainCostWidgetPresenter when chart data is undefined renders loading 1`] = `null`;

exports[`DomainCostWidgetPresenter when there are chart data renders a Doughnut 1`] = `
<div
  className="domain-legend-container"
>
  <CardLegend
    colors={
      Array [
        "#0c3d8a",
        "#859ec4",
        "#D8DADA",
        "#487ebd",
        "#007898",
        "#7fbbcb",
        "#58737a",
        "#92aab1",
        "#eee7a5",
        "#f1e364",
        "#f8b100",
        "#dcaeb9",
        "#ad405a",
        "#d9d4b9",
        "#a19351",
        "#886b22",
        "#7d7a58",
        "#564e02",
        "#c1dfc4",
        "#7dab6e",
        "#008d3a",
        "#36542c",
        "#263428",
      ]
    }
    labels={
      Array [
        "Assortment & Space Management (22412.97 EUR)",
        "B2B Experience (798.72 EUR)",
        "Checkout (Multichannel Experience) (48097.69 EUR)",
      ]
    }
  />
  <div
    className="canvas-container"
  >
    <Doughnut
      data={
        Object {
          "datasets": Array [
            Object {
              "backgroundColor": Array [
                "rgb(109,46,235)",
                "rgb(222,28,70)",
                "rgb(220,159,172)",
                "rgb(125,62,251)",
              ],
              "data": Array [
                49198.94,
                60463.31,
                983.12,
                6401.91,
              ],
            },
          ],
          "labels": Array [
            "Customer Retention(49198.94)",
            "Checkout (Multichannel Experience)(60463.31)",
            "B2B Experience(983.12)",
            "Service Management(6401.91)",
          ],
        }
      }
      options={
        Object {
          "legend": Object {
            "display": false,
          },
          "responsive": true,
        }
      }
    />
  </div>
</div>
`;

exports[`DomainCostWidgetPresenter when there are no chart data doesnt render a Doughnut 1`] = `<p />`;
